lib/cc/cli/version_checker.rb

```diff
@@ -0,0 +1,83 @@
+require "uuid"
+
+module CC
+  module CLI
+    class VersionChecker
+      include CC::CLI::Output
+
+      VERSION_CHECK_TIMEOUT = 60 * 60 # 1 Hour in seconds
+      DEFAULT_VERSIONS_URL = "https://versions.codeclimate.com".freeze
+
+      def check
+        return unless check_version?
+
+        print_new_version_message if outdated?
+      end
+
+      private
+
+      def check_version?
+        first_arg = ARGV.first
+        if %w[--check-version].include? first_arg
+          ARGV.shift
+          true
+        else
+          false
+        end
+      end
+
+      def outdated?
+        api_response["outdated"] == true
+      end
+
+      def latest_version
+        api_response["latest"]
+      end
+
+      def print_new_version_message
+        warn "A new version (v#{latest_version}) is available"
+      end
+
+      def api_response
+        @api_response ||=
+          begin
+            JSON.parse(api_response_body)
+          rescue JSON::ParserError => error
+            CLI.debug(error)
+            # We don't know so pretend all is peachy. We'll try again next time.
+            {
+              "latest" => version,
+              "outdated" => false,
+            }
+          end
+      end
+
+      def api_response_body
+        Timeout.timeout(5) do
+          if http_response.is_a? Net::HTTPSuccess
+            http_response.body
+          else
+            raise Net::HTTPFatalError.new("HTTP Error", http_response)
+          end
+        end
+      rescue Net::HTTPFatalError, Timeout::Error, Errno::EINVAL, Errno::ECONNRESET, EOFError, Errno::ECONNREFUSED => error
+        CLI.debug(error)
+        ""
+      end
+
+      def http_response
+        @http_response ||=
+          begin
+            uri = URI.parse(ENV.fetch("CODECLIMATE_VERSIONS_URL", DEFAULT_VERSIONS_URL))
+            uri.query = { version: version, uid: UUID.new.generate }.to_query
+            Net::HTTP.get_response(uri)
+          end
+      end
+
+      def version
+        @version ||=
+          File.read(File.expand_path("../../../VERSION", __dir__))
```

Comment:
Might not be worth it, but WDYT of DRYing this with https://github.com/codeclimate/codeclimate/blob/master/lib/cc/cli/version.rb somehow? I'm slightly concerned if we re-org this repo, we'll have to be careful to update multiple path references.